File: server/src/main/scala/producer/message_generator/properties_generator/PropertiesGenerator.scala
package producer.message_generator.properties_generator

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.data_generators.json_generator.JsonGenerator

case class PropertiesGenerator(
    generator: JsonGenerator
)

object PropertiesGenerator:
    def fromPb(v: pb.PropertiesGenerator): PropertiesGenerator =
        val generator = v.generator match
            case pb.PropertiesGenerator.Generator.GeneratorFromJsonObject(v) => JsonGenerator.fromPb(v)
            case _ => throw new Exception("Unknown properties generator type")

        PropertiesGenerator(
            generator = generator
        )

    def toPb(v: PropertiesGenerator): pb.PropertiesGenerator =
        val generator = v.generator match
            case v: JsonGenerator => pb.PropertiesGenerator.Generator.GeneratorFromJsonObject(JsonGenerator.toPb(v))

        pb.PropertiesGenerator(
            generator = generator
        )

File: ui/components/ui/ProducerSession/producer-task/message-generator/properties-generator/properties-generator.ts
import * as pb from '../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { JsonGenerator, jsonGeneratorFromPb, jsonGeneratorToPb } from '../data-generators/json-generator/json-generator';

export type PropertiesGenerator = {
  type: 'properties-generator',
  generator: JsonGenerator
};

export function propertiesGeneratorFromPb(v: pb.PropertiesGenerator): PropertiesGenerator {
  let generator: PropertiesGenerator['generator'];
  switch (v.getGeneratorCase()) {
    case pb.PropertiesGenerator.GeneratorCase.GENERATOR_FROM_JSON_OBJECT:
      generator = jsonGeneratorFromPb(v.getGeneratorFromJsonObject()!);
      break;
    default:
      throw new Error(`Unknown properties generator: ${v.getGeneratorCase()}`);
  }

  return { type: 'properties-generator', generator };
};

export function propertiesGeneratorToPb(v: PropertiesGenerator): pb.PropertiesGenerator {
  const message = new pb.PropertiesGenerator();
  message.setGeneratorFromJsonObject(jsonGeneratorToPb(v.generator));

  return message;
}
